### PR TITLE
[3.13] GH-124567: Revert the Incremental GC in 3.13

### PR DESCRIPTION
**Pull request by Yhg1s**
Revert the Incremental GC in 3.13.

This reverts PRs #116206 (commit 15309329b65a285cb7b3071f0f08ac964b61411b), #117120 (e28477f214276db941e715eebc8cdfb96c1207d9), #117213 (8bef34f625e21886b1c64544c060e19ee2e229bf), #117422 (ddf814db744006e0f42328aa15ace97c9d8ad681), #118313 (2ba1aed596ff9e7c5e193cf990f1f20f08bbf116), #123268 (b1372e2f1afe6bd89ffaebe7383418acd8171408) and #123395 (aca65112fe3e0d2e65a73d98186ac0309a70b65d). Does not revert the raising of the first generation threshold from 700 to 2000, nor changes to C APIs (even private ones).

* Issue: gh-124567

----
📚 Documentation preview 📚: https://cpython-previews--124770.org.readthedocs.build/

### CODE REVIEW

**Comment by Yhg1s** [comment]
This is a fairly simple revert, we'll have to address the what's new docs and news entries but I want to get this out there for review and testing in the meantime.

**Comment by AlexWaygood** [comment]
Happy to report that this cuts around 50% off the time it takes to do an HTML build of the CPython docs locally with Sphinx. On the tip of the 3.13 branch, the second script I posted in https://github.com/python/cpython/issues/124567#issue-2549191807 reports a Sphinx docs-build time of 66.28s for me locally; with this PR, it reports 34.38s.

**Comment by nascheme** [comment]
A little benchmarking from my desktop PC.  "Sphinx" is build/html for typing.rst.  The debug column is for the `--with-debug` build.  The `opt` column is for optimized build without LTO or PGO.  The "mypy" runs are for the mypy testsuite after disabling the `gc.set_threshold()` call it does (it sets a much higher threshold than default one).  The mypy run creates quite a big heap but not much cyclic garbage.  The Sphinx run is a smaller heap but quite a lot of cyclic garbage with smallish objects.


Description | debug [sec] | opt [sec] | peak rss [MB]
-- | -- | -- | --
Sphinx, 3.12 | 5.17 | 1.71 | 88.1
Sphinx, 3.13 inc | 10.25 | 1.92 | 96.9
Sphinx, 3.13 revert | 6.03 | 1.87 | 96.9
Mypy, 3.12 |   | 25.54 | 274.4
Mypy, 3.13 inc |   | 28.28 | 276.2
Mypy, 3.13 revert |   | 25.51 | 272.4

File: Modules/gcmodule.c [hunk 1]
```diff
@@ -326,7 +331,7 @@ gc_get_objects_impl(PyObject *module, Py_ssize_t generation)
     }
 
     PyInterpreterState *interp = _PyInterpreterState_GET();
-    return _PyGC_GetObjects(interp, (int)generation);
+    return _PyGC_GetObjects(interp, generation);
```
Comment (AlexWaygood) [inline Modules/gcmodule.c]: There's lots of compiler warnings on this line on Windows showing up on GitHub's "files changed" tab for this PR: 
image

**Comment by AlexWaygood** [comment]
We can also see that the `Doctest` CI job on this PR took only 4 minutes, as opposed to e.g. 16-17 minutes on other recent `3.13` PRs such as
- https://github.com/python/cpython/actions/runs/11087447714/job/30806712091?pr=124738
- https://github.com/python/cpython/actions/runs/11092139738/job/30816812499?pr=124764
- https://github.com/python/cpython/actions/runs/11084048667/job/30798940333?pr=124727.

**Review by nascheme — APPROVED** [review]
LGTM.  I compared gc.c from before GH-108362 (d5ebf8b71fd18d7a1f2) and it looks like your reversion got the right bits.

**Comment by Yhg1s** [comment]
I _believe_ the ABI changes are acceptable (the size of a private struct within another private struct changes, so later fields in the outer struct move around -- but as far as I can tell those structs are not exposed to users). The hypothesis failure seems unrelated (but possibly worth investigating).

File: Doc/library/gc.rst [hunk 2]
```diff
@@ -40,18 +40,11 @@ The :mod:`gc` module provides the following functions:
 
 .. function:: collect(generation=2)
 
-   Perform a collection.  The optional argument *generation*
+   With no arguments, run a full collection.  The optional argument *generation*
    may be an integer specifying which generation to collect (from 0 to 2).  A
-   :exc:`ValueError` is raised if the generation number is invalid. The sum of
+   :exc:`ValueError` is raised if the generation number  is invalid. The sum of
```
Comment (markshannon) [inline Doc/library/gc.rst]: Extra whitespace (if it matters)

**Review by markshannon — APPROVED** [review]
One (unimportant) whitespace regression, otherwise this looks correct to me.

**Comment by pablogsal** [comment]
> I _believe_ the ABI changes are acceptable (the size of a private struct within another private struct changes, so later fields in the outer struct move around -- but as far as I can tell those structs are not exposed to users). The hypothesis failure seems unrelated (but possibly worth investigating).

I checked the ABI changes and they are safe as this just complains about the size change of `_gc_runtime_state`. This won't affect wheels or anything using the ABI and for profilers the offsets have changed but this is not a problem if they are using the new offsets fields here:

https://github.com/python/cpython/blob/fe58596cce205c5c7aced9de3fd21767e47ee250/Include/internal/pycore_runtime.h#L193

**Review by pablogsal — APPROVED** [review]

**Comment by AA-Turner** [comment]
https://github.com/AA-Turner/cpython/commit/d7acbbd02f9e269075d98b8a30ab6c289ad91764 resolves *What's New* by just removing all references to the incremental GC.

I'm not sure there's much value discussing that the incremental GC was in pre-releases and was then reverted in What's New as it will serve to confuse the majority of readers, the detail should be in the NEWS entry.

A

File: Doc/whatsnew/3.13.rst [hunk 3]
```diff
@@ -509,19 +509,6 @@ Incremental garbage collection
 The cycle garbage collector is now incremental.
```
Comment (gpshead) [inline Doc/whatsnew/3.13.rst]: shouldn't this entire section from line 504 through the gh-108632 credit below be removed?

Comment (Yhg1s) [reply Doc/whatsnew/3.13.rst]: Done by including Adam's changes to the docs.

**Comment by ned-deily** [comment]
FWIW, a negative performance impact of the revert to some microbenchmarks: https://github.com/python/cpython/issues/124567#issuecomment-2386994636

**Pull request merged**